staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -93,6 +93,10 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	resultCh := make(chan interface{})
 	var tw timeoutWriter
 	tw, w = newTimeoutWriter(w)
+
+	// Make a copy of request and work on it in new goroutine
+	// to avoid race condition when accessing/modifying request (e.g. headers)
+	rCopy := r.Clone(r.Context())
```

Comment:
I'm not a fan of this approach - in general I would expect the headers to not be modified and if some handler really has to do that for whever reason, I would expect the copying happening in those places.

Can we rather try closing the request in places that actually do modify the headers?

Reply:
It seems that @aojea mentioned similar concern in the PR:
https://github.com/kubernetes/kubernetes/pull/108455#issuecomment-1059408390

It also seems that @liggitt was the one who lgtm-ed the original PR that was removing the header - Jordan, can you please shed some more light on it if there is any additional context to keep in mind here?

Reply:
We *really* do not want anything in the server other than the authentication filter, or in downstream proxy backends seeing or using the authorization header. That can leak credentials or lead code to trust credentials that have not been validated properly.

Reply:
@liggitt - I got this; the question is - whether you agree it would be more intuitive if instad authentication filter was actually doing a deep-copy if it deletes the  header

Reply:
mmm... not sure... 99% of the filters and handling chain reasonably expect to be operating in a single thread. the timeout filter is sort of the odd one out, it's certainly easier to reason about safety/correctness if it is the one to do special handling

Reply:
I agree, in my opinion the issue arises because expect that their filters/handling code will be running in one goroutine (which is quite reasonable expectations). They are probably not aware that in the middle of the chain we split processing into new goroutine. I think that this should be responsibility of this special filter to protect other parts of the code from race condition if it decides to launch new goroutine.

Reply:
do we know how other libraries are dealing with this situation?

Reply:
Interesting solution using a context.WithValue to pass the mutex

https://github.com/kubernetes/kubernetes/pull/108716

Reply:
This is an interesting option, but it still puts responsibility on the person writing filter/handler to remember to use it (if we only pass mutex in context, and don't actually wrap headers in some struct that guarantees thread safety). I believe that its timeout handler responsibility to ensure that filters/handlers around it can work properly assuming single threaded flow of execution. @liggitt @wojtek-t @aojea  I think we should accept one solution and implement it. I won't oppose if majority votes for other solution than presented in this pr.

Reply:
OK - I'm not huge fan of it, but I can buy the argument it's "least ugly" option.